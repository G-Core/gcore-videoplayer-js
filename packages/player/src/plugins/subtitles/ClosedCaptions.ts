import { Events, UICorePlugin, Browser, template, $ } from '@clappr/core'
import { reportError, trace } from '@gcorevideo/utils'
import assert from 'assert'

import { CLAPPR_VERSION } from '../../build.js'
import type { TextTrackItem } from '../../internal.types.js'

import '../../../assets/subtitles/style.scss'
import subtitlesOffIcon from '../../../assets/icons/new/subtitles-off.svg'
import subtitlesOnIcon from '../../../assets/icons/new/subtitles-on.svg'
import comboboxHTML from '../../../assets/subtitles/combobox.ejs'
import stringHTML from '../../../assets/subtitles/string.ejs'

import { isFullscreen } from '../utils/fullscreen.js'
import type { ZeptoResult } from '../../types.js'
import { ExtendedEvents } from '../media-control/MediaControl.js'

const VERSION: string = '2.19.14'

const LOCAL_STORAGE_CC_ID = 'gplayer.plugins.cc.selected'

const T = 'plugins.cc'

/**
 * Configuration options for the {@link ClosedCaptions} plugin.
 * @beta
 */
export type ClosedCaptionsPluginSettings = {
  /**
   * Initially selected subtitles language.
   */
  language?: string
}

/**
 * `PLUGIN` that provides a UI to select the subtitles when available.
 * @beta
 *
 * @remarks
 * The plugin is activated when closed captions tracks are detected in the media source.
 * It shows a familiar "CC" button with a dropdown menu to select the subtitles language.
 *
 * Depends on:
 *
 * - {@link MediaControl}
 *
 * Configuration options -  {@link ClosedCaptionsPluginSettings}
 * @example
 * ```ts
 * import { ClosedCaptions } from '@gcorevideo/player'
 *
 * Player.registerPlugin(ClosedCaptions)
 *
 * new Player({
 *   ...
 *   cc: {
 *     language: 'pt-BR',
 *   },
 * })
 * ```
 */
export class ClosedCaptions extends UICorePlugin {
  private isPreselectedApplied = false

  private active = false

  private open = false

  private track: TextTrackItem | null = null

  private tracks: TextTrackItem[] = []

  private $line: ZeptoResult | null = null

  /**
   * @internal
   */
  get name() {
    return 'cc'
  }

  /**
   * @internal
   */
  get supportedVersion() {
    return { min: CLAPPR_VERSION }
  }

  /**
   * @internal
   */
  static get version() {
    return VERSION
  }

  private static readonly templateControl = template(comboboxHTML)

  private static readonly templateLine = template(stringHTML)

  /**
   * @internal
   */
  override get attributes() {
    return {
      class: 'media-control-cc media-control-dd__wrap',
    }
  }

  /**
   * @internal
   */
  override get events() {
    return {
      'click #gplayer-cc-menu [data-item]': 'onItemSelect',
      'click #gplayer-cc-button': 'toggleMenu',
    }
  }

  private get preselectedLanguage(): string {
    return (
      this.core.options.cc?.language ??
      this.core.options.subtitles?.language ??
      ''
    )
  }

  /**
   * @internal
   */
  override bindEvents() {
    this.listenTo(this.core, Events.CORE_READY, this.onCoreReady)
    this.listenTo(this.core, Events.CORE_RESIZE, this.playerResize)
    this.listenTo(
      this.core,
      Events.CORE_ACTIVE_CONTAINER_CHANGED,
      this.onContainerChanged,
    )
  }

  private onCoreReady() {
    const mediaControl = this.core.getPlugin('media_control')
    assert(mediaControl, 'media_control plugin is required')
    this.listenTo(mediaControl, Events.MEDIACONTROL_RENDERED, this.mount)
    this.listenTo(mediaControl, Events.MEDIACONTROL_HIDE, () => {
      this.hideMenu()
    })
    this.listenTo(
      mediaControl,
      ExtendedEvents.MEDIACONTROL_MENU_COLLAPSE,
      (from: string) => {
        if (from !== this.name) {
          this.hideMenu()
        }
      },
    )
  }

  private onContainerChanged() {
    this.listenTo(
      this.core.activeContainer,
      Events.CONTAINER_FULLSCREEN,
      this.playerResize,
    )
    this.listenTo(
      this.core.activeContainer,
      'container:advertisement:start',
      this.onStartAd,
    )
    this.listenTo(
      this.core.activePlayback,
      Events.PLAYBACK_SUBTITLE_AVAILABLE,
      this.onSubtitleAvailable,
    )
    this.listenTo(
      this.core.activePlayback,
      Events.PLAYBACK_SUBTITLE_CHANGED,
      this.onSubtitleChanged,
    )
    this.listenTo(this.core.activeContainer, Events.CONTAINER_CLICK, () => {
      // TODO test
      this.hideMenu()
    })

    // fix for iOS
    const video = this.core.activePlayback.el
    assert(video, 'video element is required')

    video.addEventListener('webkitbeginfullscreen', () => {
      if (Browser.isiOS) {
        video.classList.add('ios-fullscreen')
      }
    })

    video.addEventListener('webkitendfullscreen', () => {
      if (Browser.isiOS) {
        video.classList.remove('ios-fullscreen')
      }
    })
  }

  private onSubtitleAvailable() {
    trace(`${T} onSubtitleAvailable`)
    this.applyTracks()
    this.mount()
  }

  private onSubtitleChanged({ id }: { id: number }) {
    trace(`${T} onSubtitleChanged`, { id })
    if (id === -1) {
      this.clearSubtitleText()
    }
    for (const track of this.tracks) {
      if (track.id === id) {
        track.track.mode = 'showing'

        this.setSubtitleText(this.getSubtitleText(track.track))

        track.track.oncuechange = (e) => {
          try {
            if (track.track.activeCues?.length) {
              const html = (track.track.activeCues[0] as VTTCue).getCueAsHTML()

              this.setSubtitleText(html)
            } else {
              this.clearSubtitleText()
            }
          } catch (error) {
            reportError(error)
          }
        }
      } else {
        track.track.oncuechange = null
        track.track.mode = 'hidden'
      }
    }
  }

  private applyTracks() {
    try {
      this.tracks = this.core.activePlayback.closedCaptionsTracks
      this.applyPreselectedSubtitles()
      this.render()
    } catch (error) {
      reportError(error)
    }
  }

  private onStartAd() {
    if (this.active && this.core.activeContainer) {
      this.hide()
      this.listenTo(
        this.core.activeContainer,
        'container:advertisement:finish',
        this.onFinishAd,
      )
    }
  }

  private onFinishAd() {
    this.show()
    this.stopListening(
      this.core.activeContainer,
      'container:advertisement:finish',
      this.onFinishAd,
    )
  }

  private playerResize() {
    trace(`${T} playerResize`)
    const shouldShow =
      this.core.activeContainer &&
      isFullscreen(this.core.activeContainer.el) &&
      this.track &&
      this.track.track.mode &&
      Browser.isiOS &&
      this.active

    if (shouldShow) {
      this.show()
    }

    try {
      this.resizeFont()
      this.clampPopup()
    } catch (error) {
      reportError(error)
    }
  }

  /**
   * Hides the subtitles menu and the subtitles.
   */
  hide() {
    this.active = false
    this.open = false
    this.renderIcon()
    this.$el.find('#gplayer-cc-menu').hide()
    this.$el.find('#gplayer-cc-button').attr('aria-expanded', 'false')
    this.$line.hide()
    if (this.tracks) {
      for (const t of this.tracks) {
        t.track.mode = 'hidden'
      }
    }
  }

  /**
   * Shows the subtitles menu and the subtitles.
   */
  show() {
    this.active = true
    this.renderIcon()
    if (
      this.core.activeContainer &&
      isFullscreen(this.core.activeContainer.el) &&
      this.track &&
      this.track.track.mode &&
      Browser.isiOS
    ) {
      this.$line.hide()
      this.track.track.mode = 'showing'
    } else {
      this.$line.show()
    }
  }

  private shouldRender() {
    return this.tracks?.length > 0
  }

  private resizeFont() {
    if (!this.$line) {
      return
    }

    const skinWidth = this.core.activeContainer.$el.width()

    this.$line.find('p').css('font-size', skinWidth * 0.03)
  }

  /**
   * @internal
   */
  override render() {
    if (!this.core.activeContainer) {
      return this
    }

    this.$el.html(
      ClosedCaptions.templateControl({
        tracks: this.tracks ?? [],
        i18n: this.core.i18n,
        current: this.track?.id ?? -1,
      }),
    )
    this.$el.find('#gplayer-cc-menu').hide()
    this.open = false
    this.core.activeContainer.$el.find('#gplayer-cc-line').remove()
    this.$line = $(ClosedCaptions.templateLine())
    this.resizeFont()
    this.clampPopup()

    this.core.activeContainer.$el.append(this.$line)

    this.updateSelection()

    this.renderIcon()

    return this
  }

  private findById(id: number) {
    return this.tracks.find((track) => track.id === id) ?? null
  }

  private selectItem(item: TextTrackItem | null) {
    this.clearSubtitleText()
    this.track = item

    this.updateSelection()
  }

  private onItemSelect(event: MouseEvent) {
    // event.target does not exist for some reason in tests
    const id =
      ((event.target ?? event.currentTarget) as HTMLElement).dataset?.item ??
      '-1'

    localStorage.setItem(LOCAL_STORAGE_CC_ID, id) // TODO store language instead
    this.selectItem(this.findById(Number(id)))
    this.hideMenu()
    return false
  }

  private applyPreselectedSubtitles() {
    if (!this.isPreselectedApplied) {
      this.isPreselectedApplied = true
      if (!this.preselectedLanguage) {
        return
      }
      setTimeout(() => {
        this.selectItem(
          this.tracks.find(
            (t) => t.track.language === this.preselectedLanguage,
          ) ?? null,
        )
      }, 300) // TODO why delay?
    }
  }

  private hideMenu() {
    this.open = false
    this.$el.find('#gplayer-cc-menu').hide()
    this.$el.find('#gplayer-cc-button').attr('aria-expanded', 'false')
  }

  private toggleMenu() {
    this.core
      .getPlugin('media_control')
      .trigger(ExtendedEvents.MEDIACONTROL_MENU_COLLAPSE, this.name)
    this.open = !this.open
    if (this.open) {
      this.$el.find('#gplayer-cc-menu').show()
    } else {
      this.$el.find('#gplayer-cc-menu').hide()
    }
    this.$el.find('#gplayer-cc-button').attr('aria-expanded', this.open)
  }

  private itemElement(id: number): ZeptoResult {
    // TODO fix semantically
    return this.$el.find(`#gplayer-cc-menu [data-item="${id}"]`).parent()
  }

  private allItemElements(): ZeptoResult {
    return this.$el.find('#gplayer-cc-menu li') // TODO fix semantically
  }

  private selectSubtitles() {
    const trackId = this.track ? this.track.id : -1

    this.core.activePlayback.closedCaptionsTrackId = trackId // TODO test
  }

  private getSubtitleText(track: TextTrack) {
    const currentTime = this.core.activePlayback?.getCurrentTime() ?? 0
    const cues = track.cues
    const lines = []

    if (cues && cues.length) {
      for (const cue of cues) {
        if (currentTime >= cue.startTime && currentTime <= cue.endTime) {
          lines.push((cue as VTTCue).getCueAsHTML().textContent)
          // TODO break?
        }
      }
    }

    return lines.join('\n')
  }

  private setSubtitleText(text: string | DocumentFragment) {
    this.$line.find('p').html(text)
  }

  private clearSubtitleText() {
    this.setSubtitleText('')
  }

  private updateSelection() {
    if (!this.track) {
      this.hide()
    } else {
      this.show()
    }
    this.selectSubtitles()
    this.highlightCurrentSubtitles()
  }

  private highlightCurrentSubtitles() {
    this.allItemElements()
      .removeClass('current')
      .find('a')
      .removeClass('gcore-skin-active')
      .attr('aria-checked', 'false')

    const currentLevelElement = this.itemElement(
      this.track ? this.track.id : -1,
    )
    currentLevelElement
      .addClass('current')
      .find('a')
      .addClass('gcore-skin-active')
      .attr('aria-checked', 'true')
  }

  private renderIcon() {
    // render both icons at once
    const icon = this.active ? subtitlesOnIcon : subtitlesOffIcon
    this.$el.find('#gplayer-cc-button').html(icon)
  }

  private clampPopup() {
    const availableHeight = this.core
      .getPlugin('media_control')
      .getAvailablePopupHeight()
    this.$el.find('#gplayer-cc-menu').css('max-height', `${availableHeight}px`)
  }

  private mount() {
    if (this.shouldRender()) {
      const mediaControl = this.core.getPlugin('media_control')
      mediaControl.slot('cc', this.$el)
    }
  }
}
