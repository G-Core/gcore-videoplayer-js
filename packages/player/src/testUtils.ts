import { $, UICorePlugin } from '@clappr/core'
import Events from 'eventemitter3'
import { vi } from 'vitest'

export function createMockCore(
  options: Record<string, unknown> = {},
  container: any = createMockContainer(options),
) {
  const el = document.createElement('div')
  const emitter = new Events()
  return Object.assign(emitter, {
    el,
    $el: $(el),
    activePlayback: container.playback,
    activeContainer: container,
    i18n: {
      t: vi.fn().mockImplementation((key: string) => key),
    },
    options: {
      ...options,
    },
    configure: vi.fn(),
    getPlaybackType: vi.fn(),
    getPlugin: vi.fn(),
    load: vi.fn(),
    trigger: emitter.emit,
  })
}

export function createMockPlugin() {
  return Object.assign(new Events(), {
    enable: vi.fn(),
    disable: vi.fn(),
  })
}

export function createSpinnerPlugin() {
  return Object.assign(createMockPlugin(), {
    show: vi.fn(),
    hide: vi.fn(),
  })
}

export function createMockPlayback(name = 'mock') {
  const emitter = new Events()
  return Object.assign(emitter, {
    name,
    currentLevel: -1,
    el: document.createElement('video'),
    dvrEnabled: false,
    dvrInUse: false,
    isAudioOnly: false,
    levels: [],
    consent: vi.fn(),
    play: vi.fn(),
    pause: vi.fn(),
    stop: vi.fn(),
    destroy: vi.fn(),
    seek: vi.fn(),
    seekPercentage: vi.fn(),
    getDuration: vi.fn().mockImplementation(() => 100),
    enterPiP: vi.fn(),
    exitPiP: vi.fn(),
    getPlaybackType: vi.fn(),
    getStartTimeOffset: vi.fn().mockImplementation(() => 0),
    getCurrentTime: vi.fn().mockImplementation(() => 0),
    isHighDefinitionInUse: vi.fn().mockImplementation(() => false),
    mute: vi.fn(),
    unmute: vi.fn(),
    volume: vi.fn(),
    configure: vi.fn(),
    attemptAutoPlay: vi.fn().mockImplementation(() => true),
    canAutoPlay: vi.fn().mockImplementation(() => true),
    onResize: vi.fn().mockImplementation(() => true),
    setPlaybackRate: vi.fn(),
    switchAudioTrack: vi.fn(),
    trigger: emitter.emit,
  })
}

export function createMockContainer(
  options: Record<string, unknown> = {},
  playback: any = createMockPlayback(),
) {
  const el = playback.el
  const emitter = new Events()
  return Object.assign(emitter, {
    el,
    playback,
    options: {
      ...options,
    },
    $el: $(el),
    disableMediaControl: vi.fn(),
    enableMediaControl: vi.fn(),
    enterPiP: vi.fn(),
    exitPiP: vi.fn(),
    getDuration: vi.fn().mockReturnValue(0),
    getPlugin: vi.fn(),
    getPlaybackType: vi.fn(),
    getStartTimeOffset: vi.fn().mockReturnValue(0),
    isDvrInUse: vi.fn().mockReturnValue(false),
    isDvrEnabled: vi.fn().mockReturnValue(false),
    isHighDefinitionInUse: vi.fn().mockReturnValue(false),
    isPlaying: vi.fn().mockReturnValue(false),
    play: vi.fn(),
    seek: vi.fn(),
    seekPercentage: vi.fn(),
    switchAudioTrack: vi.fn(),
    trigger: emitter.emit,
  })
}

export function createMockMediaControl(core: any) {
  const mediaControl = new UICorePlugin(core)
  // TODO <div class="media-control-layer">
  mediaControl.$el.html(
    `<div class="media-control-left-panel" data-media-control></div>
    <div class="media-control-right-panel" data-media-control></div>
    <div class="media-control-center-panel" data-media-control></div>`,
  )
  // @ts-ignore
  mediaControl.mount = vi.fn()
  // @ts-ignore
  mediaControl.container = core.activeContainer
  // @ts-ignore
  mediaControl.toggleElement = vi.fn()
  vi.spyOn(mediaControl, 'trigger')
  core.$el.append(mediaControl.$el)
  return mediaControl
}

export function createMockBottomGear(core: any) {
  const plugin: any = new UICorePlugin(core)
  plugin.getItem = vi.fn()
  plugin.addItem = vi.fn().mockImplementation((name: string, $el: any) => {
    const existing = plugin.$el.find(`[data-${name}]`)
    if (existing.length) {
      return existing
    }
    return $('<li></li>')
      .attr(`data-${name}`, '')
      .append($el)
      .appendTo(plugin.$el)
  })
  plugin.refresh = vi.fn()
  return plugin
}
