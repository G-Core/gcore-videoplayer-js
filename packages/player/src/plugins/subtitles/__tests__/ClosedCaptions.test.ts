import { beforeEach, describe, expect, it, vi } from 'vitest'

import { ClosedCaptions } from '../ClosedCaptions.js'
import { createMockCore, createMockMediaControl } from '../../../testUtils.js'
import { ExtendedEvents } from '../../media-control/MediaControl.js'

import { Events } from '@clappr/core'

// import { LogTracer, Logger, setTracer } from '@gcorevideo/utils'

// Logger.enable('*')
// setTracer(new LogTracer('ClosedCaptions.test'))

describe('ClosedCaptions', () => {
  let core: any
  let mediaControl: any
  let cc: ClosedCaptions
  beforeEach(() => {
    core = createMockCore()
    mediaControl = createMockMediaControl(core)
    mediaControl.getAvailablePopupHeight = vi.fn().mockReturnValue(211)
    core.getPlugin = vi.fn().mockImplementation((name) => {
      if (name === 'media_control') {
        return mediaControl
      }
      return null
    })
    cc = new ClosedCaptions(core)
  })
  describe('basically', () => {
    beforeEach(() => {
      core.emit(Events.CORE_READY)
      core.activePlayback.el = document.createElement('video')
      core.emit(Events.CORE_ACTIVE_CONTAINER_CHANGED, core.activeContainer)
    })
    it('should render', () => {
      expect(cc.el.innerHTML).toMatchSnapshot()
      expect(cc.$el.find('#gplayer-cc-button').length).toEqual(1)
      expect(cc.$el.find('#gplayer-cc-menu').css('display')).toEqual('none')
      expect(core.activeContainer.$el.find('#gplayer-cc-line').length).toEqual(
        1,
      )
    })
    it('should clamp popup to availableheight', () => {
      expect(cc.$el.find('#gplayer-cc-menu').css('max-height')).toBe('211px')
    })
    describe('until subtitle tracks are available', () => {
      it('should not mount', () => {
        expect(mediaControl.slot).not.toHaveBeenCalledWith(
          'cc',
          expect.anything(),
        )
        emitSubtitleAvailable(core)
        expect(mediaControl.slot).toHaveBeenCalledWith('cc', cc.$el)
      })
    })
    describe('when viewport is resized', () => {
      beforeEach(() => {
        mediaControl.getAvailablePopupHeight = vi.fn().mockReturnValue(197)
        core.activeContainer.emit(Events.CONTAINER_RESIZE, {
          width: 320,
          height: 260,
        })
        core.emit(Events.CORE_RESIZE, { width: 320, height: 260 })
      })
      it('should clamp popup height', () => {
        expect(cc.$el.find('#gplayer-cc-menu').css('max-height')).toBe('197px')
      })
    })
    describe('when media control is rerendered', () => {
      beforeEach(() => {
        emitSubtitleAvailable(core)
        mediaControl.trigger(Events.MEDIACONTROL_RENDERED)
      })
      it('should mount', () => {
        expect(mediaControl.slot).toHaveBeenCalledWith('cc', cc.$el)
      })
    })
    describe('when button is clicked', () => {
      beforeEach(() => {
        emitSubtitleAvailable(core)
        cc.$el.find('#gplayer-cc-button').click()
      })
      it('should open menu', () => {
        expect(cc.$el.find('#gplayer-cc-menu').css('display')).not.toBe('none')
        expect(cc.$el.find('#gplayer-cc-button').attr('aria-expanded')).toBe(
          'true',
        )
      })
      it('should collapse all other menus', () => {
        expect(mediaControl.trigger).toHaveBeenCalledWith(
          ExtendedEvents.MEDIACONTROL_MENU_COLLAPSE,
          'cc',
        )
      })
    })
    describe('when clicked twice', () => {
      beforeEach(() => {
        emitSubtitleAvailable(core)
        cc.$el.find('#gplayer-cc-button').click().click()
      })
      it('should collapse the menu', () => {
        expect(cc.$el.find('#gplayer-cc-menu').css('display')).toBe('none')
        expect(cc.$el.find('#gplayer-cc-button').attr('aria-expanded')).toBe(
          'false',
        )
      })
    })
    describe('when media control is hidden', () => {
      beforeEach(() => {
        emitSubtitleAvailable(core)
        cc.$el.find('#cc-button').click()
        mediaControl.trigger(Events.MEDIACONTROL_HIDE)
      })
      it('should hide menu', () => {
        expect(cc.$el.find('#gplayer-cc-menu').css('display')).toBe('none')
      })
    })
    describe('when container is clicked', () => {
      beforeEach(() => {
        emitSubtitleAvailable(core)
        cc.$el.find('#gplayer-cc-button').click()
        core.activeContainer.emit(Events.CONTAINER_CLICK)
      })
      it('should hide menu', () => {
        expect(cc.$el.find('#gplayer-cc-menu').css('display')).toBe('none')
        expect(cc.$el.find('#gplayer-cc-button').attr('aria-expanded')).toBe(
          'false',
        )
      })
    })
    describe('when subtitle is changed', () => {
      beforeEach(async () => {
        emitSubtitleAvailable(core)
        await new Promise((resolve) => setTimeout(resolve, 100))
        core.activePlayback.getCurrentTime = vi.fn().mockReturnValue(7)
        core.activeContainer.getCurrentTime = vi.fn().mockReturnValue(7)
        core.activePlayback.closedCaptionsTracks[1].track.cues = [
          {
            startTime: 0,
            endTime: 5,
            text: 'Alright this is me',
            getCueAsHTML: vi
              .fn()
              .mockImplementation(() =>
                document.createTextNode('Alright this is me'),
              ),
          },
          {
            startTime: 5,
            endTime: 10,
            text: 'Welcome to my channel',
            getCueAsHTML: vi
              .fn()
              .mockImplementation(() =>
                document.createTextNode('Welcome to my channel'),
              ),
          },
          {
            startTime: 10,
            endTime: 15,
            text: 'Thank you for watching',
            getCueAsHTML: vi
              .fn()
              .mockImplementation(() =>
                document.createTextNode('Thank you for watching'),
              ),
          },
        ]
        core.activePlayback.emit(Events.PLAYBACK_SUBTITLE_CHANGED, { id: 2 })
      })
      it('should activate subtitle track', () => {
        expect(core.activePlayback.closedCaptionsTracks[1].track.mode).toBe(
          'showing',
        )
        expect(core.activePlayback.closedCaptionsTracks[0].track.mode).toBe(
          'hidden',
        )
      })
      it('should show active subtitle text', () => {
        expect(
          core.activeContainer.$el.find('#gplayer-cc-line').text().trim(),
        ).toEqual('Welcome to my channel')
      })
    })
    describe('when subtitle is selected from menu', () => {
      beforeEach(() => {
        emitSubtitleAvailable(core)
        cc.$el.find('#gplayer-cc-menu li:nth-child(2) a').click()
      })
      it('should activate subtitle track', () => {
        expect(core.activePlayback.closedCaptionsTrackId).toEqual(2)
      })
      it('should highlight selected menu item', () => {
        expect(
          cc.$el.find('#gplayer-cc-menu li:nth-child(2)').hasClass('current'),
        ).toBe(true)
        expect(
          cc.$el
            .find('#gplayer-cc-menu li:nth-child(2) a')
            .attr('aria-checked'),
        ).toBe('true')
        expect(cc.$el.find('#gplayer-cc-menu li.current').length).toBe(1)
        expect(
          cc.$el.find('#gplayer-cc-menu li a[aria-checked="true"]').length,
        ).toBe(1)
      })
      it('should collapse menu', () => {
        expect(cc.$el.find('#gplayer-cc-menu').css('display')).toBe('none')
      })
    })
    describe('when user turns off subtitles', () => {
      beforeEach(() => {
        emitSubtitleAvailable(core)
        core.activePlayback.closedCaptionsTracks[1].track.mode = 'showing'
        core.activePlayback.closedCaptionsTrackId = 2
        core.activePlayback.emit(Events.PLAYBACK_SUBTITLE_CHANGED, { id: 2 })
        cc.$el.find('#gplayer-cc-button').click()
        cc.$el.find('#gplayer-cc-menu li:nth-child(3) a').click() // off
      })
      it('should hide menu', () => {
        expect(cc.$el.find('#gplayer-cc-menu').css('display')).toBe('none')
      })
      it('should hide subtitle text', () => {
        expect(
          core.activeContainer.$el.find('#gplayer-cc-line').text().trim(),
        ).toBe('')
      })
      it('should deactivate subtitle track', () => {
        expect(core.activePlayback.closedCaptionsTrackId).toEqual(-1)
        expect(core.activePlayback.closedCaptionsTracks[0].track.mode).toBe(
          'hidden',
        )
        expect(core.activePlayback.closedCaptionsTracks[1].track.mode).toBe(
          'hidden',
        )
      })
      it('should select menu item', () => {
        expect(
          cc.$el.find('#gplayer-cc-menu li:nth-child(3)').hasClass('current'),
        ).toBe(true)
        expect(
          cc.$el
            .find('#gplayer-cc-menu li:nth-child(3) a')
            .attr('aria-checked'),
        ).toBe('true')
        expect(cc.$el.find('#gplayer-cc-menu li.current').length).toBe(1)
        expect(
          cc.$el.find('#gplayer-cc-menu li a[aria-checked="true"]').length,
        ).toBe(1)
      })
    })
  })
})

function emitSubtitleAvailable(core: any) {
  core.activePlayback.closedCaptionsTracks = [
    {
      id: 1,
      name: 'English',
      track: {
        language: 'en',
        kind: 'subtitles',
        label: 'English',
        mode: 'hidden',
        cues: [],
      },
    },
    {
      id: 2,
      name: 'Spanish',
      track: {
        language: 'es',
        kind: 'subtitles',
        label: 'Spanish',
        mode: 'hidden',
        cues: [],
      },
    },
  ]
  core.activePlayback.emit(Events.PLAYBACK_SUBTITLE_AVAILABLE)
  core.activeContainer.emit(Events.CONTAINER_SUBTITLE_AVAILABLE)
}
