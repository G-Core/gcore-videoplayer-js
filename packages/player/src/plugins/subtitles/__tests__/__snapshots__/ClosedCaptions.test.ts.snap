// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ClosedCaptions > basically > should render 1`] = `
"<button class="media-control-button media-control-icon gcore-skin-button-color media-control-dd" id="gplayer-cc-button" aria-haspopup="menu" aria-expanded="false" aria-controls="gplayer-cc-menu">/assets/icons/new/subtitles-off.svg</button>

<ul class="gcore-skin-bg-color media-control-dd__popup" id="gplayer-cc-menu" role="menu" style="display: none; max-height: 211px;">
    
    <li class="current">
        <a href="#" class="gcore-skin-text-color gcore-skin-active" data-item="-1" role="menuitemradio" aria-checked="true">
            off
        </a>
    </li>
</ul>"
`;
